WIP: MGMT-299 and MGMT-368 documentation (#31697)
openshift/openshift-docs: modules/mgmt-power-remediation-baremetal-about.adoc

// Module included in the following assemblies:

// * machine_management/mgmt-power-remediation-baremetal

[id="mgmt-power-remediation-baremetal-about_{context}"]
= About power-based remediation of bare metal
In a bare metal cluster, remediation of nodes is critical to ensure the overall health of the cluster. Physically remediating a cluster can be challenging and any delay in putting the machine to a safe or an operational state increases the time the cluster remains in a degraded state, and the risk that subsequent failures might bring the cluster offline. Power-based remediation helps counter such challenges.

Instead of reprovisioning the nodes, power-based remediation uses a power controller to power off an inoperable node. This type of remediation is also called power fencing.

> The OpenShift Container Platform uses the `MachineHealthCheck` controller to detect faulty bare metal nodes by initiating power-cycling. Power-based remediation is fast and reboots unhealthy hosts instead of removing them from the cluster.

Reviewer: s/The OpenShift Container Platform/{product-title}

Power-based remediation meets the following objectives:

* Allows the recovery of control plane nodes
* Prevents data loss in hyperconverged environments.
* Minimizes downtime for stateful applications that run on the unhealthy nodes.

[id="machine-health-checks-bare-metal_{context}"]
== MachineHealthChecks on bare metal

Machine deletion on bare metal cluster triggers reprovisioning of a bare metal host.
Usually bare metal reprovisioning is a lengthy process, during which the cluster
is missing compute resources and applications might be interrupted.
To change the default remediation process from machine deletion to host power-cycle,

> annotate the MachineHealthCheck resource with the

Reviewer: `MachineHealthCheck` resource 

per https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#object-references

The remediation process operates as follows:

. The MachineHealthCheck (MHC) controller detects that a node is unhealthy.
. The MHC notifies the controller in the cluster API provider for managed bare metal hardware (CAPM3) controller.
. The controller in the cluster API provider for managed bare metal hardware (CAPM3) requests to power-off the unhealthy node.
. After the power is off, the Node CR is deleted, which allows the cluster to reschedule the affected workload on other nodes.
. The cluster API provider for managed bare metal hardware (CAPM3) requests to power on the node.
. After the node is up, the node re-registers itself with the cluster, causing the creation of a new Node CR
. After  the Node CR is recreated, the cluster API provider for managed bare metal hardware (CAPM3) restores the annotations and labels that existed on the unhealthy node before its deletion.

[NOTE]
====
If the power operations did not complete, the cluster API provider for managed bare metal hardware (CAPM3) triggers the reprovisioning of the unhealthy node.
====

[id="mgmt-creating-mhc-baremetal_{context}"]
== Creating a MachineHealthCheck resource for bare metal

.Prerequisites

> * The OpenShift Container Platform is installed using installer-provisioned infrastructure (IPI).

Reviewer: s/The OpenShift Container Platform/{product-title}

* Access to Baseboard Management Controller (BMC) credentials (or BMC access to each node)
* Network access to the BMC interface of the unhealthy node.

.Procedure
. Create a `healthcheck.yml` file that contains the definition of your machine health check.
. Apply the `healthcheck.yml` file to your cluster using the following command:

> [source,terminal]

Reviewer: [source,terminal]
    ----
    $ oc apply -f healthcheck.yml
    ----

// Module included in the following assemblies:

// * machine_management/mgmt-power-remediation-baremetal

[id="mgmt-power-remediation-baremetal-about_{context}"]
= About power-based remediation of bare metal
In a bare metal cluster, remediation of nodes is critical to ensuring the overall health of the cluster. Physically remediating a cluster can be challenging and any delay in putting the machine into a safe or an operational state increases the time the cluster remains in a degraded state, and the risk that subsequent failures might bring the cluster offline. Power-based remediation helps counter such challenges.

Instead of reprovisioning the nodes, power-based remediation uses a power controller to power off an inoperable node. This type of remediation is also called power fencing.

> {product-title} uses the `MachineHealthCheck` controller to detect faulty bare metal nodes by initiating power-cycling. Power-based remediation is fast and reboots unhealthy hosts instead of removing them from the cluster.

Reviewer 2: `MachineHealthCheck` controller doesn't detect faulty nodes by initiating power-cycling.

it detect faulty nodes in the same way for all platforms. power-cycling is the remediation action taken in BM.

Author: Rephrased to: 
{product-title} uses the `MachineHealthCheck` controller to detect faulty bare metal nodes. Power-based remediation is fast and reboots faulty nodes instead of removing them from the cluster.

Power-based remediation provides the following capabilities:

* Allows the recovery of control plane nodes
* Reduces the risk data loss in hyperconverged environments
* Reduces the downtime associated with recovering physical machines

[id="machine-health-checks-bare-metal_{context}"]
== MachineHealthChecks on bare metal

Machine deletion on bare metal cluster triggers reprovisioning of a bare metal host.
Usually bare metal reprovisioning is a lengthy process, during which the cluster
is missing compute resources and applications might be interrupted.
To change the default remediation process from machine deletion to host power-cycle,
annotate the `MachineHealthCheck` resource with the
`machine.openshift.io/remediation-strategy: external-baremetal` annotation.

After you set the annotation, unhealthy machines are power-cycled by using
BMC credentials.

[id="mgmt-understanding-remediation-process_{context}"]
== Understanding the remediation process

The remediation process operates as follows:

. The MachineHealthCheck (MHC) controller detects that a node is unhealthy.

> . The MHC notifies the bare metal machine controller requests to power-off the unhealthy node.

Reviewer 2: Something is slightly off here?
maybe:
The MHC notifies the bare metal machine controller **which** requests to power-off the unhealthy node.

. After the power is off, the node is deleted, which allows the cluster to reschedule the affected workload on other nodes.
. The bare metal machine controller requests to power on the node.
. After the node is up, the node re-registers itself with the cluster, resulting in the creation of a new node.
. After the node is recreated, the bare metal machine controller restores the annotations and labels that existed on the unhealthy node before its deletion.

[NOTE]
====

> If the power operations did not complete, the bare metal machine controller triggers the reprovisioning of the unhealthy node unless this is a master node or a node that is reprovisioned externally.

Reviewer 2: ```suggestion
If the power operations did not complete, the bare metal machine controller triggers the reprovisioning of the unhealthy node unless this is a master node or a node that was provisioned externally.
```

// Module included in the following assemblies:

// * machine_management/mgmt-power-remediation-baremetal

[id="mgmt-power-remediation-baremetal-about_{context}"]
= About power-based remediation of bare metal
In a bare metal cluster, remediation of nodes is critical to ensure the overall health of the cluster. Physically remediating a cluster can be challenging and any delay in putting the machine to a safe or an operational state increases the time the cluster remains in a degraded state, and the risk that subsequent failures might bring the cluster offline. Power-based remediation helps counter such challenges.

Instead of reprovisioning the nodes, power-based remediation uses a power controller to power off an inoperable node. This type of remediation is also called power fencing.

The {product-title} uses the `MachineHealthCheck` controller to detect faulty bare metal nodes by initiating power-cycling. Power-based remediation is fast and reboots unhealthy hosts instead of removing them from the cluster.

> Power-based remediation meets the following objectives:

Reviewer 3: ```suggestion
Power-based remediation provides the following capabilities:
```

* Allows the recovery of control plane nodes

> * Prevents data loss in hyperconverged environments.

Reviewer 3: If the node was left running, then in theory it could be generating or processing data that isn't being synchronised to the rest of the cluster.  Which would count as data loss.

Likewise if we recover a master, the cluster can survive the subsequent loss of another one - allowing incoming data to be process and not lost.

So technically true, but possibly a little misleading.

```suggestion
* Reduces the risk data loss in hyperconverged environments.
* Reduces the downtime associated with recovering physical machines
```

* Minimizes downtime for stateful applications that run on the unhealthy nodes.

[id="machine-health-checks-bare-metal_{context}"]
== MachineHealthChecks on bare metal

Machine deletion on bare metal cluster triggers reprovisioning of a bare metal host.
Usually bare metal reprovisioning is a lengthy process, during which the cluster
is missing compute resources and applications might be interrupted.
To change the default remediation process from machine deletion to host power-cycle,
annotate the `MachineHealthCheck` resource with the

> `machine.openshift.io/remediation-strategy: external-baremetal` annotation.

Reviewer 3: Note that this part of the procedure will change in 4.9

Author: @aireilly There is a sample YAML file in the topic that shows the use of this annotation

After you set the annotation, unhealthy machines are power-cycled by using
BMC credentials.

[id="mgmt-understanding-remediation-process_{context}"]
== Understanding the remediation process

The remediation process operates as follows:

. The MachineHealthCheck (MHC) controller detects that a node is unhealthy.

> . The MHC notifies the controller in the cluster API provider for managed bare metal hardware (CAPM3) controller.

Reviewer 2: in openshift we use CAPBM (not CAPM3)

. The controller in the cluster API provider for managed bare metal hardware (CAPM3) requests to power-off the unhealthy node.

> . After the power is off, the Node CR is deleted, which allows the cluster to reschedule the affected workload on other nodes.

Reviewer 2: s/CR/resource as `node` is not a *custom* resource

. The cluster API provider for managed bare metal hardware (CAPM3) requests to power on the node.

> . After the node is up, the node re-registers itself with the cluster, causing the creation of a new Node CR

Reviewer 2: same here for CR

Reviewer 3: I think that was my fault, but yes "Node resource" would be more technically correct

> . After  the Node CR is recreated, the cluster API provider for managed bare metal hardware (CAPM3) restores the annotations and labels that existed on the unhealthy node before its deletion.

Reviewer 2: double space,

[NOTE]
====

> If the power operations did not complete, the cluster API provider for managed bare metal hardware (CAPM3) triggers the reprovisioning of the unhealthy node.

Reviewer 2: unless this is a master node or node that can not be reprovisioned (e.g. `ExternallyProvisioned`)

====

[id="mgmt-creating-mhc-baremetal_{context}"]
== Creating a MachineHealthCheck resource for bare metal

.Prerequisites

* The {product-title} is installed using installer-provisioned infrastructure (IPI).
* Access to Baseboard Management Controller (BMC) credentials (or BMC access to each node)
* Network access to the BMC interface of the unhealthy node.

.Procedure
. Create a `healthcheck.yml` file that contains the definition of your machine health check.
. Apply the `healthcheck.yml` file to your cluster using the following command:

[source,terminal]
----

> $ oc apply -f healthcheck.yml

Author: Could not find any cue on the yml vs yaml, but since most instances in the docs use `.yaml` or YAML, so changed it to `healthcheck.yaml`

kind: MachineHealthCheck
metadata:
  name: example <1>
  namespace: openshift-machine-api
  annotations:
    machine.openshift.io/remediation-strategy: external-baremetal <2>
spec:
  selector:
    matchLabels:
      machine.openshift.io/cluster-api-machine-role: <role> <3>
      machine.openshift.io/cluster-api-machine-type: <role> <3>
      machine.openshift.io/cluster-api-machineset: <cluster_name>-<label>-<zone> <4>
  unhealthyConditions:
  - type:    "Ready"
    timeout: "300s" <5>
    status: "False"
  - type:    "Ready"
    timeout: "300s" <5>
    status: "Unknown"
  maxUnhealthy: "40%" <6>
  nodeStartupTimeout: "10m" <7>
----

<1> Specify the name of the machine health check to deploy.
<2> For bare metal clusters, you must include the `machine.openshift.io/remediation-strategy: external-baremetal` annotation in the `annotations` section to enable power-cycle remediation. With this remediation strategy, unhealthy hosts are rebooted instead of removed from the cluster.
<3> Specify a label for the machine pool that you want to check.
<4> Specify the machine set to track in `<cluster_name>-<label>-<zone>` format. For example, `prod-node-us-east-1a`.

> <5> Specify the timeout duration for a node condition. If a condition is met for the duration of the timeout, the machine will be remediated. Long timeouts can result in long periods of downtime for a workload on an unhealthy machine.

Reviewer 4: ```suggestion
<5> Specify the timeout duration for the node condition. If the condition is met for the duration of the timeout, the machine will be remediated. Long timeouts can result in long periods of downtime for a workload on an unhealthy machine.
```
